1. Add RestClient and JsonApiClient

kashizui opened the pull request:
Factor out general methods from BundleServiceClient into RestClient and
add a new JsonApiClient to make calls to JSON API endpoints.

@klopyrev

kashizui committed:
Add RestClient and JsonApiClient

Factor out general methods from BundleServiceClient into RestClient and
add a new JsonApiClient to make calls to JSON API endpoints.

klopyrev commented on codalab/client/rest_client.py:
```diff
@@ -0,0 +1,94 @@
+    def __init__(self, base_url):
+        self._base_url = base_url
+
+    def _get_access_token(self):
```
Add a comment that the implementation should be thread-safe.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+from codalab.common import http_error_to_exception
+
+
+# TODO(sckoo): deal with more error cases
```
Please do this part properly. Exception handling is important. See for example my implementation of wrap_exception.

kashizui replied:
@klopyrev Should I make a `JsonApiException` like you did with `BundleServiceException`, or should I try to translate errors to the standard CodaLab errors in `codalab.common`?

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+        self.id_ = id_
+
+
+class JsonApiClient(RestClient, BundleClient):
```
Why is this a BundleClient?

kashizui replied:
I'm mixing in BundleClient in order inherit its `login` method, in order to fit in with how CodalabManager authenticates a client.

klopyrev replied:
I don't understand this logic. Please leave it out of the pull request and bring it in once you needed it. I'd like to see the code that uses it.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+    Simple JSON API client.
+    """
+    def __init__(self, address, auth_handler, authenticate):
+        self.address = address
```
Not used.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+    """
+    def __init__(self, address, auth_handler, authenticate):
+        self.address = address
+        self.auth_handler = auth_handler
```
Not used.

klopyrev replied:
I shouldn't have to track these down. Please pay more attention to writing clean code.

kashizui replied:
These are used by the CodalabManager to authenticate, I'll add comments to clarify.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+
+    @wrap_exception
+    def delete(self, resource_type, resource_id, params=None):
+        if isinstance(resource_id, list):
```
I have no idea why you're supporting both ways of deleting things.

kashizui replied:
You're right it would probably best would just be to always do multiple-delete. I haven't implemented bulk-delete for some of the API yet so I will need to make sure to do that in the next PRs.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+    return result
+
+
+def pack_document(objects, type_):
```
This packing and unpacking of documents seems like premature optimization... Just my two cents.

kashizui replied:
I use every feature I implemented here in the subsequent PRs. I kind of built this up incrementally as I needed it.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+        :return:
+        """
+        if isinstance(data, list):
+            path = self._get_resource_path(resource_type)
```
What use case does this fit?

kashizui replied:
`cl chown` updates the owner of potentially multiple bundles, and so this allows a batch update. But as you pointed out with delete, I can simplify this down to always assuming batch operations.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+                linkage = relationship['data']
+                if linkage is None:
+                    obj[key] = None
+                elif linkage['id'] in included[linkage['type']]:
```
This is copied code.

kashizui replied:
I'll look into refactoring it.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+                        obj[key].append({'id': linkage['id']})
+            else:
+                linkage = relationship['data']
+                if linkage is None:
```
Use same control flow structure as below for 'data':

data = relationship['data']
if isinstance(..., list):
elif instance(..., dict):
elif None??

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+            if isinstance(relationship['data'], list):
+                obj[key] = []
+                for linkage in relationship['data']:
+                    if linkage['id'] in included[linkage['type']].keys():
```
No need for .keys

kashizui replied:
Good catch!

klopyrev replied:
You're relying on defaultdict behavior here... That's kind of weird. I'd use a tuple: type, id

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,299 @@
+                    if linkage['id'] in included[linkage['type']].keys():
+                        obj[key].append(unpack_object(included[linkage['type']][linkage['id']]))
+                    else:
+                        obj[key].append({'id': linkage['id']})
```
Why no type in this case? Just include the whole object.

kashizui commented:
@klopyrev Addressed all comments

kashizui commented:
Found a bug: this causes the worker to throw this error:

```
Traceback (most recent call last):
  File "dev/codalab-cli/worker/main.py", line 15, in <module>
    from bundle_service_client import BundleServiceClient
  File "/Users/sckoo/Developer/codalab-cli/worker/bundle_service_client.py", line 13, in <module>
    from codalab.client.rest_client import RestClient, RestClientException
ImportError: No module named codalab.client.rest_client
```

will have to address this tomorrow.

klopyrev commented:
Yeah, the worker can't depend on the codalab libraries.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,369 @@
+            except urllib2.HTTPError as e:
+                # Translate known errors to the standard CodaLab errors
+                exc = http_error_to_exception(e.code, e.read())
+                if isinstance(exc, UsageError):
```
Why just this type? What about permissions errors? Why do we even need to translate?

klopyrev replied:
Add some more comments if you keep it this way.

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,369 @@
+        try:
+            data = document.get('data', None)
+            if isinstance(data, list):
+                result = [unpack_object(d) for d in document['data']]
```
s/document['data']/data/

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,369 @@
+            if isinstance(data, list):
+                result = [unpack_object(d) for d in document['data']]
+            elif isinstance(data, dict):
+                result = unpack_object(document['data'])
```
s/document['data']/data/

klopyrev commented on codalab/client/json_api_client.py:
```diff
@@ -0,0 +1,369 @@
+                path=self._get_resource_path(resource_type),
+                query_params=self._pack_params(params),
+                data=self._pack_document(data, resource_type)))
+        return result if isinstance(data, list) else result[0]
```
This line seems broken.

klopyrev commented on codalab/client/rest_client.py:
```diff
@@ -0,0 +1,95 @@
+
+            # Read the response.
+            response = conn.getresponse()
+            if response.status != 200:
```
This will not get properly translated into a UsageError, even though it has an HTTP error code and it should.

kashizui commented:
@klopyrev Finished addressing comments
- Moved RestClient into worker directory
- `RestClient._upload_with_chunked_encoding` throws `HTTPError`
- Added comments clarifying the "translation" of errors
- `s/document['data']/data`
- Clarified confusing line in `JsonApiClient.update`

klopyrev commented:
LGTM
